rfcs/69-editor-design-system.md

```diff
@@ -0,0 +1,297 @@
+  - Orange was chosen for Code, since we use Rust; 🦀
+  - Emerald was chosen for Resources as a toss-up due to sheer elimination of available hues. If resources turn out to not be significant enough to warrant their own color, we might choose to re-assign it to something else.
+- **Pure Red, Green and Blue** — These colors are used to denote _themselves_ (e.g. in material/shader inputs) as well as to denote the X, Y and Z 3D axes. The strong, well established association of these colors from other tools is leveraged to help users orient themselves in the 3D space. Importantly, their intensity values are tweaked independently so that they're all _perceptually_ at the same brightness level.
+- **Primary and Destructive** — These pastel shades are intended for primary buttons in modal dialogs, and for destructive actions. (Such as deleting an item)
+
+#### Typography
+
+TBA
+
+#### Iconography
+
+Our icon style is heavily influenced by Blender's, but further simplified, with the primary goal being to make icon creation as economical as possible. Specifically, a single icon should _not_ take more than **10 minutes** to be created and fully fleshed out by someone comfortable with a vector tool.
+
+Icons are:
+
+- Vector-based;
+- Authored in three separate sizes, (16x16, 24x24, 32x32) with small tweaks and simplifications made where appropriate to accomodate for the smaller sizes;
+- Hinted to the pixel grid;
+- Drawn with a 1px outline style, with a few exceptions (such as folders) meant to be clear visual landmarks;
+- Either single color, or at most two colors, with one being the foreground color;
+- Whenever “shading” is needed, dotted lines or some other dithering/meshing pattern should be used;
+
+![Sample Bevy Editor Design System Icons](assets/69-editor-design-system/icons.png)
+
+![Same Icon at Different Sizes](assets/69-editor-design-system/icon-sizes.png)
+
+### Standard Widgets
+
+#### Button
+
+TBA
+
+#### Checkbox
+
+TBA
+
+#### Radio Button
+
+TBA
+
+#### ScrollView
+
+TBA
+
+#### TextInput
+
+TBA
+
+#### ColorInput
+
+TBA
+
+#### NumericInput
+
+TBA
+
+#### SelectInput
+
+TBA
+
+#### Slider
+
+TBA
+
+
+####
+
+## Part 3 — Implementation
+
+### The `bevy_editor_ui` Crate
+
+A new `bevy_editor_ui` crate is introduced, with the purpose of housing the reusable standard widgets built on top of `bevy_ui`, colors, typography styles, icons and other assets that make up the Bevy Editor Design System.
+
+This crate is intended to be consumed both by a future `bevy_editor` crate, as well as by unofficial Bevy Editor plugins, so that they can both match the visual style of the editor and tightly integrate with it.
+
+### Standard Widget Patterns
+
+#### Bundle Structure
+
+Standard widgets from `bevy_editor_ui` adhere to a standard bundle structure. For a standard widget named `MyWidget`:
+
+| Component | Required? |  Role  | Access By Widget's Systems | Access By Widget's User  | Reactive?
```

Comment:
> This entire table is unreviewable until we have a clear idea (at least) about the UI architecture we want, which we don't. We're still exploring patterns of UI and ECS integration and I've yet to see something compelling on paper, let alone an actual RFC about it.

If unreviewable, I can definitely leave it out of the RFC once it's out of draft. But IMO, the _Editor_ UI architecture can afford to be far more opinionated and comprehensive than the general UI architecture we expose in Bevy UI, since it will be consumed primarily internally from within the same repo; the churn of upgrading in case we change our minds is much smaller. (We can even specifically flag it as experimental, so that people don't build third party plugins for the editor initially expecting upgrading will be easy)

> - Data being owned by the widget excludes an entire class of designs (like Druid's, were widgets borrow and mutate data in place)

Hmm. Haven't looked into Druid yet, so I can't say I fully understand what a borrowing/mutation-first taken on a UI framework actually entails, but I don't see owning data as mutually exclusive with tapping into data from somewhere else via mutable or immutable references. The owned data could be a _predicate_ that allows indirectly controlling borrowing and mutating data in place, like a search filter closure for a query, an `Entity` or maybe even a reflected path into some component.

For example: in my proof or concept, I'm using a unit struct plus a trait with an associated type to display arbitrary entities with a specific component as items in a tree view, but the tree view filter (for searching) will be owned by the component as local state.

> - The write-only widget out excludes another entire class of design implementation (React's single source of truth were data is mutated in-place and diff-ed for changes)

I got the idea for this specific pattern _from_ React, as it's analogous to the `forwardRef()`/`useImperativeHandle()` mechanism. React has two "ways" of handling things like form inputs and other interactive elements: controlled and uncontrolled. 

- Controlled has the parent be the source of truth, and requires all changes to be first propagated from the child to the parent via events  (e.g `onChange`). The parent event handler issues a local state update, and the parent "re-renders", then the child is reconciled and also "re-renders". This is the pattern you are likely referring to;
- Uncontrolled has the child be the source of truth, so changes are applied internally as local state updates. The parent can then use the child's exposed imperative handle (obtained via a ref) to read the data only when needed. (e.g. when submitting a form)

Both preserve the single source of truth for each piece of information, the difference is just where it ends up being kept. The same codebase will often up using both patterns for different things, depending on whether one needs to react to all changes granularly or just read a snapshot after a specific action. (Controlled is very boilerplate-y, and has potential performance implications if the parent is too complex)

Note: React also exposes _methods_ in this imperative handle, but I think we might be better served by something else like an event-based mechanism for this.